design/multiple-listeners-design.md

```diff
@@ -0,0 +1,87 @@
+Contour's main purupose is still a Layer 7 Ingress Controller, not a Layer 4 load balancer.
+
+## Goals
+- Allow adding multiple (an arbitrary number) of extra listeners to Contour's Envoy fleet
+- Maintain Contour's current insecure to secure redirection by default
+
+## Non Goals
+- TCP Layer 4 load balancing support on any address
+- UDP Layer 4 load balancing support
+
+## High-Level Design
+
+### Configuring the secure and insecure ports
+
+Currently, Contour allows the configuration of the secure and insecure listening ports for Envoy using the `--envoy-service-https-port` and the `--envoy-service-http-port` parameters respectively.
+(It's fair to say that these parameters are confusingly named).
+
+These should be exposed in the configuration file as and `secureListenerPort` and `insecureListenerPort` respectively.
+They should also have aliases added for the command line flags to `--envoy-secure-listener-port` and `--envoy-insecure-listener-port` respectively.
+
+The Contour Operator may then add support to its `Contour` CRD for tweaking this value.
+
+For the redirect to be able to work currently, the example YAML and the Operator currently have the Envoy listen on port `8443` and `8080` respectively *inside its Pod*, but this is translated out to `80` and `443` by the `Service` or `Type: Loadbalancer` included there.
+This means that the redirect generated can redirect from `http://somedomanin.com` to `https://somedomain.com` and have everything work.
+
+For this feature to work more broadly, we must add a config parameter `IncludePortOnRedirect`, which, if set to `true`, will have the HTTP -> HTTPS redirect go from `http://somedomain.com/` to `https://omedomain.com:<secureport>/`.
+This will resolve [#1300](https://github.com/projectcontour/contour/issues/1300) as well, and improve support for non-standard default HTTPS ports.
+Note that this is only required if you change the **default** port, that gets the redirect to secure.
+If you add additional HTTPS listener ports, they will **not** get a redirection.
+
+### HTTPProxy processing
+
+For HTTPProxy resources, we will add an optional `port` field to the `virtualhost` YAML stanza.
+
+The rules for how this `port` field interacts with the rest of Contour would be this:
+- no port set
+    - TLS details set - default secure port, redirect will be created for you
+    - no TLS details set - default insecure port, same `permitInsecure` behavior.
+- port set
+    - TLS details set - any port other than the insecure or secure ports, allow. No redirection created for you.
+    - no TLS details set - any port other than the insecure or secure ports, allow.
+    - use the insecure or the secure port - fatal error, HTTPProxy will not be processed.
+
+The documents for the `port` field must clearly state that it is used for specifying a non-default port *only*. Don't use it for port 80 and 443.
```

Comment:
What's the rationale for this? Note that it can create new error cases (proxy is OK, then contour config changed and proxy is now broken)

Reply:
The idea was just to make it easier to write the handling for the "supplied a port" case.

I'll make this clearer, but the port that is specified in the HTTPProxy *must* be the port on the **`Service`**, not the port on the Pod. This is one of the things I need to address in the detailed design - Contour needs to know what port to put in URLs, and we currently don't distinguish that from "what port to tell Envoy to bind to".

In the example YAML, we tell Envoy to bind to `8080` and `8443` (to avoid requiring the container to be run as root), but assume that those will be translated to `80` and `443` respectively by the Service. We just haven't documented this requirement really.

Part of changing this is that we need to make the requirements explicit somehow.